File: doc/haskell/GeometryLow/RealModulus.hs
module GeometryLow.RealModulus where

mod0Real, modMinusPlusHalfReal :: Float -> Float -> Float
mod0Real m x
    | x < 0     = mod0Real m (x + m)
    | x < m     = x
    | otherwise = mod0Real m (x - m)
modMinusPlusHalfReal m x
    | x <= -m/2 = mod0Real m (x + m)
    | x <=  m/2 = x
    | otherwise = mod0Real m (x - m)

mod_0_360 :: Float -> Float
mod_0_360 = mod0Real 360

isConvex, isConcave :: Float -> Bool
isConvex angle = mod_0_360 angle <= 180
isConcave = not . isConvex

areConvex, areConcave :: Float -> Float -> Bool
areConvex angle1 angle2 = isConvex (angle2 - angle1)
areConcave angle1 angle2 = not $ areConvex angle1 angle2

File: app/assets/math/haskell/app/RealModulus.js
//module Number.RealModulus where

//mod0Real, modMinusPlusHalfReal :: Float -> Float -> Float

function mod0Real(m, x)
{
	while (x < 0 || x >= m) x+= -Math.sign(x)*m;
	return x;
}
function modMinusPlusHalfReal(m, x)
{
	while (x <= -m/2 || x > m/2) x+= -Math.sign(x)*m;
	return x;
}

//mod_0_360 :: Float -> Float
function mod_0_360(x) {return mod0Real(360, x);}
const mod_0_180 = x => mod0Real(180, x);

//isConvex, isConcave :: Float -> Bool
function isConvex (angle) {return mod_0_360(angle) <= 180;}
function isConcave(angle) {return !isConvex(angle);}

//areConvex, areConcave :: Float -> Float -> Bool
function areConvex (angle1, angle2) {return isConvex  (angle2 - angle1);}
function areConcave(angle1, angle2) {return !areConvex(angle1,  angle2);}
